src/ray/gcs/gcs_server/gcs_actor_scheduler.h

```diff
@@ -0,0 +1,260 @@
+  void Schedule(std::shared_ptr<GcsActor> actor);
+
+  /// Set the failed handler of scheduling.
+  ///
+  /// \param handler Invoked when there are no available nodes to schedule the actor.
+  void SetScheduleFailureHandler(
+      std::function<void(std::shared_ptr<GcsActor>)> &&handler) {
+    RAY_CHECK(handler);
+    schedule_failed_handler_ = std::move(handler);
+  }
+
+  /// Set the success handler of scheduling.
+  ///
+  /// \param handler Invoked when the actor is created on the worker successfully.
+  void SetScheduleSuccessHandler(
+      std::function<void(std::shared_ptr<GcsActor>)> &&handler) {
+    RAY_CHECK(handler);
+    schedule_success_handler_ = std::move(handler);
+  }
+
+  /// Cancel the scheduling actors associated with the specified node id.
+  ///
+  /// \param node_id ID of the node where the worker is located.
+  /// \return Actors associated with the specified node id.
+  std::vector<std::shared_ptr<GcsActor>> CancelOnNode(const ClientID &node_id);
+
+  /// Cancel the scheduling actor associated with the specified node and worker.
+  ///
+  /// \param node_id ID of the node where the worker is located.
+  /// \param worker_id ID of the worker that the actor is creating on.
+  /// \return The actor associated with the specified node id and worker id.
+  std::shared_ptr<GcsActor> CancelOnWorker(const ClientID &node_id,
+                                           const WorkerID &worker_id);
+
+ protected:
+  /// Lease a worker from the specified node for the specified actor.
+  ///
+  /// \param actor This object has the resources needed to lease workers from the
+  /// specified node.
+  /// \param node The node that the worker will be leased from.
+  virtual void LeaseWorkerFromNode(std::shared_ptr<GcsActor> actor,
+                                   std::shared_ptr<rpc::GcsNodeInfo> node);
+  /// Handler to process `WorkerLeasedReply`.
+  ///
+  /// \param actor Contains the resources needed to lease workers from the specified node.
+  /// \param reply The reply of `RequestWorkerLeaseRequest`.
+  void HandleWorkerLeasedReply(std::shared_ptr<GcsActor> actor,
+                               const rpc::RequestWorkerLeaseReply &reply);
+
+  /// Create the specified actor on the specified worker.
+  ///
+  /// \param actor The actor to be created.
+  /// \param worker The worker that the actor will created on.
+  void CreateActorOnWorker(std::shared_ptr<GcsActor> actor,
+                           std::shared_ptr<GcsLeasedWorker> worker);
+
+  /// Select a node from alive nodes randomly.
+  std::shared_ptr<rpc::GcsNodeInfo> SelectNodeRandomly() const;
+
+  /// Get or create a node client by the specified node info.
+  ///
+  /// \param node_id ID of the specified node.
+  /// \param node The node info which contains the node address.
+  /// \return New node client if there is no cached client related to the node id, else
+  /// the cached one.
+  std::shared_ptr<rpc::NodeManagerWorkerClient> GetOrCreateNodeClient(
```

Comment:
hi, @stephanie-wang. Did you mean having `GcsActorScheduler` inherit from `WorkerLeaseInterface`? 
I discussed with @wumuzi520 about this. Doing this doesn't seem to have benefit. Because:
1) `GcsActorScheduler` is different from `RayletClient`, as it only leases workers for actor creation tasks and it as some GCS-specific logic. So they can't be used interchangeably. 
2) The `ReturnWorker` method is not needed by `GcsActorScheduler`, because the leased workers are for actors and won't be returned. 
3) Regarding tests, we can just mock the `LeaseWorkerFromNode` method to avoid real RPC. 
4) IIUIC, the current 2-step approach (leasing a worker from ralet and then pushing the actor creation task to worker) might be a temporary approach. In the future, we may want to optimize it to only one step (just push the task to raylet, and let raylet push the task to a worker) to simplify the process.  

Please let us know if we missed something.

Reply:
I just meant to use `WorkerLeaseInterface` instead of `NodeManagerWorkerClient`.